src/core/AutoRest.Core/Validation/RecursiveObjectValidator.cs

```diff
@@ -1,85 +1,77 @@
-                        }
-                    }
-                }
-                else if (entity.GetType().IsClass && entity.GetType() != typeof(string))
+                return Enumerable.Empty<ValidationMessage>();
+            }
+
+            // ensure that the rules can be re-enumerated without re-evaluating the enumeration.
+            var collectionRules = rules.ReEnumerable();
+
+            var list = entity as IList;
+            if (list != null)
+            {
+                // Recursively validate each list item and add the 
+                // item index to the location of each validation message
+                return list.SelectMany((item, index)
+                    => RecursiveValidate(item, collectionRules).Select(each
+                        => each.AppendToPath($"[{index}]")));
+            }
+
+            var dictionary = entity as IDictionary;
+            if (dictionary != null)
+            {
+                if (!dictionary.IsValidatableDictionary())
                 {
-                    // Validate objects by running class rules against the object and recursively against properties
-                    foreach(var exception in ValidateObjectValue(entity, collectionRules))
-                    {
-                        yield return exception;
-                    }
-                    foreach(var exception in ValidateObjectProperties(entity))
-                    {
-                        yield return exception;
-                    }
+                    return Enumerable.Empty<ValidationMessage>();
                 }
+
+                // Recursively validate each dictionary entry and add the entry 
+                // key to the location of each validation message
+                return dictionary.SelectMany((key, value)
+                    => RecursiveValidate(value, collectionRules).Select(each
+                        => each.AppendToPath((string) key)));
             }
-            yield break;
+
+            // if this is a class, validate it's value and it's properties.
+            if (entity.GetType().IsClass && entity.GetType() != typeof(string))
+            {
+                // return the messages for both this value and its properties.
+                return ValidateObjectValue(entity, collectionRules).Concat(ValidateObjectProperties(entity));
+            }
+
+            // validate just the value.
+            return ValidateObjectValue(entity, collectionRules);
         }
 
-        private IEnumerable<ValidationMessage> ValidateObjectValue(object entity, IEnumerable<RuleAttribute> collectionRules)
+        private IEnumerable<ValidationMessage> ValidateObjectValue(object entity,
+            IEnumerable<Rule> collectionRules)
         {
             // Get any rules defined for the class of the entity
-            var classRules = entity.GetType().GetCustomAttributes<RuleAttribute>(true);
+            var classRules = entity.GetType().GetValidationRules();
```

Comment:
I pulled thru the `Rule`s themselves instead of holding onto the `RuleAttribute` because this way in the future we can instantiate more rules from other sources (like external annotation) without having to wrap them.